### PR TITLE
Seq608 location report

### PR DESCRIPTION
**Pull request by andrewsparkes**
Added the location report functionality for ticket 608 for Meeta.

### CODE REVIEW

File: app/controllers/location_reports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# Please refer to the LICENSE and README files for information on licensing and
+# authorship of this file.
+# Copyright (C) 2018 Genome Research Ltd.
+
+class LocationReportsController < ApplicationController
+  before_action :login_required
+
+  def index
+    @studies                = Study.alphabetical.pluck(:name, :id)
+    @plate_purposes         = PlatePurpose.pluck(:name, :id).sort
```
Comment (andrewsparkes) [inline app/controllers/location_reports_controller.rb]: Add include SharedBehaviour::Named to PlatePurpose, and use alphabetical scope instead.

File: app/controllers/location_reports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @location_reports       = LocationReport.order(id: :desc).page(params[:page])
+    @location_report        = LocationReport.new
```
Comment (JamesGlover) [inline app/controllers/location_reports_controller.rb]: Suggest:
LocationReport.new(user: @current_user)

File: app/controllers/location_reports_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @location_report        = LocationReport.new(location_report_parameters)
+    @location_report.user   = @current_user
+
+    if @location_report.save
+      flash[:notice] = 'Your report has been requested and will be listed at the bottom of this page when complete.'
+      redirect_to location_reports_path
+    else
+      error_messages          = @location_report.errors.full_messages.join('; ')
+      flash[:error]           = "Failed to create report: #{error_messages}"
+      @studies                = Study.alphabetical.pluck(:name, :id)
```
Comment (JamesGlover) [inline app/controllers/location_reports_controller.rb]: flash.now[:error]

File: app/controllers/location_reports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      @plate_purposes         = PlatePurpose.pluck(:name, :id).sort
+      @location_reports       = LocationReport.order(id: :desc).page(params[:page])
+      @location_report.errors.clear
```
Comment (JamesGlover) [inline app/controllers/location_reports_controller.rb]: Consider using before_action for shared lookups

File: app/models/location_report.rb [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  attr_accessor :barcodes_list
+
+  def barcodes_text
+    barcodes_list.join(' ') unless @barcodes_list.nil?
+  end
+
+  def barcodes_text=(value)
+    self.barcodes_list = if value.nil?
+                           nil
+                         else
```
Comment (JamesGlover) [inline app/models/location_report.rb]: [] rather than nil

File: app/models/location_report.rb [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                           value.strip.split(/[\s]+/)
+                         end
+  end
+
+  self.per_page = 10
```
Comment (JamesGlover) [inline app/models/location_report.rb]: https://ssg-confluence.internal.sanger.ac.uk/display/PSD/Style+guide

File: app/models/location_report.rb [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+  def headers
+    %w[Barcode HumanBarcode Type Created Location Service Study Owner]
+  end
+
+  def generate!
+    ActiveRecord::Base.transaction do
+      csv_options = { row_sep: "\r\n", force_quotes: true }
+      filename    = "location_report_#{name}.csv"
+      Tempfile.open(filename) do |tempfile|
```
Comment (JamesGlover) [inline app/models/location_report.rb]: http://ruby-doc.org/stdlib-2.4.1/libdoc/csv/rdoc/CSV.html
Using a tempfile is probably going to be necessary still, but you might be able to wrap it in the CSV library.
Not sure whet the re-opening is about. Suggest getting rid of it and making sure everything still passes/works as expected.

File: app/models/location_report.rb [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+  serialize :plate_purpose_ids, Array
+
+  before_validation :generate_name, on: %i[create build]
+
+  after_create :schedule_report
+
+  validates :report_type, presence: true
+
+  validates :study_id, numericality: true, if: :report_type_selection?, allow_nil: true
```
Comment (JamesGlover) [inline app/models/location_report.rb]: validates :study, presence: true, if: :report_type_selection?

File: app/models/location_report.rb [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  validates :start_date, presence: { if: :report_type_selection? }
+  validates :end_date, presence: { if: :report_type_selection?, date: { after_or_equal_to: :start_date } }
+
+  validates :barcodes_text, presence: { if: :report_type_barcodes? }
+  validate :barcodes_are_recognised, if: :report_type_barcodes?
+  validate :plates_are_found, if: :report_type_selection?
+
+  def report_type_selection?
```
Comment (JamesGlover) [inline app/models/location_report.rb]: http://api.rubyonrails.org/classes/ActiveRecord/Enum.html

File: app/models/location_report.rb [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  end
+
+  def report_type_barcodes?
+    report_type == 'barcodes'
+  end
+
+  def barcodes_are_recognised
+    if barcodes_list.blank?
+      errors.add(:barcodes_text, 'Please enter some barcodes in the text area')
+      return
```
Comment (JamesGlover) [inline app/models/location_report.rb]: http://guides.rubyonrails.org/i18n.html
But this has already been rested above, so probably don't need to add the error again. YOu might even be able to tweak the if: above to only run if the text is filled.

File: app/models/location_report.rb [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        invalid_barcodes_list.push(cur_bc)
+      end
+    end
+
+    unless invalid_barcodes_list.size.zero?
+      errors.add(:barcodes_text, "Invalid barcodes found: #{invalid_barcodes_list.join(',')}")
+      return
+    end
+
+    if ean13_barcodes_list.size.zero?
```
Comment (JamesGlover) [inline app/models/location_report.rb]: This check is redundant

File: app/models/location_report.rb [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+      errors.add(:barcodes_text, 'Please enter some valid barcodes (human or scannable) in the text area')
+      return
+    end
+
+    self.barcodes = ean13_barcodes_list.join(' ')
```
Comment (JamesGlover) [inline app/models/location_report.rb]: Probably serialize an array. A bit more overhead, but stops you needing to split again later.

File: app/models/plate.rb [hunk 13]
```diff
@@ -373,6 +373,30 @@ def with(subinclude)
...
 
+  def self.search_for_plates(study_id: nil, plate_purpose_ids: nil, start_date: nil, end_date: nil)
+    with_study_id(study_id)
+      .with_plate_purpose_ids(plate_purpose_ids)
+      .created_after(start_date)
+      .created_before(end_date)
+      .distinct
+  end
+
+  scope :with_study_id, (proc { |study_id|
```
Comment (JamesGlover) [inline app/models/plate.rb]: Try swithing to lambda syntax: eg:

->(variable) { block_code }

variable optional. Can drop brackets entirely is its not present.

File: app/models/location_report.rb [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+  def plates_are_found
+    unless Plate.search_for_plates(
+      study_id:             study_id,
+      start_date:           start_date,
+      end_date:             end_date,
+      plate_purpose_ids:    plate_purpose_ids
+    ).any?
+      errors.add(:study_id, 'Those selection criteria return no plates')
+    end
```
Comment (JamesGlover) [inline app/models/location_report.rb]: Error messages have the attribute name stuck to the begining.

File: app/models/location_report.rb [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+  end
+
+  def schedule_report
+    Delayed::Job.enqueue LocationReportJob.new(id)
+  end
+
+  def generate_report_rows
```
Comment (JamesGlover) [inline app/models/location_report.rb]: Try and split up a bit.

File: app/models/location_report.rb [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        if barcode_is_human_readable?(cur_bc)
+          # human readable format
+          ean13_barcodes_list.push(Barcode.human_to_machine_barcode(cur_bc))
+        elsif barcode_is_ean13?(cur_bc)
+          # ean 13 format so unchanged
+          ean13_barcodes_list.push(cur_bc)
+        end
+      end
+
+      # fetch the plates for the batch of barcodes
```
Comment (JamesGlover) [inline app/models/location_report.rb]: MOst of this block is unecessary

File: app/models/location_report.rb [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    end
+
+    if plates_list.empty?
+      yield(['Error: plates list was empty when generating the report'])
+      return
+    end
+
+    yield headers
+
+    plates_list.each do |cur_plate|
```
Comment (JamesGlover) [inline app/models/location_report.rb]: Pull into separate method, pass block through with &block (add as final argument to outer method)

File: app/models/location_report.rb [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+      if cur_plate.studies.present?
+        cur_plate.studies.each do |cur_study|
+          yield([
+            cur_plate.machine_barcode,
```
Comment (JamesGlover) [inline app/models/location_report.rb]: pull array generation into method, handle missing study. (Possibly just using something like study&.name || 'Unknown'

File: app/views/location_reports/_filter_criteria.html.erb [hunk 19]
```diff
@@ -0,0 +1,17 @@
+<%# This file is part of SEQUENCESCAPE; it is distributed under the terms of
```
Comment (JamesGlover) [inline app/views/location_reports/_filter_criteria.html.erb]: Delete

File: app/views/location_reports/_location_plate_selection_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    <%= f.label 'plate_purpose_ids', 'Plate purposes' %>
+    <%= f.select :plate_purpose_ids, options_for_select(plate_purposes, @location_report.plate_purpose_ids), { prompt: 'Select plate purposes...', include_hidden: false }, { class: 'form-control select2', :multiple => true } %>
+  </div>
+
+  <%= f.hidden_field :report_type, value: 'selection' %>
+
+  <div class='form-group actions-group'>
+    <%= f.submit('Create report from selection', autocomplete: 'off',class: 'disable_on_submit btn btn-success') %>
+  </div>
+<% end %>
```
Comment (JamesGlover) [inline app/views/location_reports/_location_plate_selection_form.html.erb]: Return at end

File: app/views/location_reports/show.html.erb [hunk 21]
```diff
@@ -0,0 +1,13 @@
+<%# This file is part of SEQUENCESCAPE; it is distributed under the terms of
```
Comment (JamesGlover) [inline app/views/location_reports/show.html.erb]: This page can be deleted

File: app/models/location_report.rb [hunk 22]
```diff
@@ -5,188 +5,156 @@
...
+    return %w[Unknown Unknown] if cur_study.blank?
+
+    cols = [] << cur_study.name ||= 'Unknown'
+    cols << if cur_study.study_metadata.present?
+              cur_study.study_metadata.faculty_sponsor.name ||= 'Unknown'
+            else
+              'Unknown'
+            end
+  end
+
```
Comment (JamesGlover) [inline app/models/location_report.rb]: def generate_study_cols_for_row(cur_study)
    return %w[Unknown Unknown] if cur_study.blank?

    cols = [] << cur_study.name || cur_study.id
    cols << cur_study.study_metadata.faculty_sponsor&.name || 'Unknown'
  end

File: config/locales/en.yml [hunk 23]
```diff
@@ -202,3 +202,10 @@ en:
     exclude_existing:
       yes: New samples
       no: All samples
+  location_reports:
+    success: "Your report has been requested and will be listed at the bottom of this page when complete."
+    errors:
+      no_rows_found: " - That selection returns no plates, no report generated."
+      invalid_barcodes_found: " - Invalid barcodes found, no report generated: "
```
Comment (JamesGlover) [inline config/locales/en.yml]: Remove -

File: spec/models/location_report_spec.rb [hunk 24]
```diff
@@ -57,21 +57,13 @@
   let(:plt_3_purpose)         { plate_3.plate_purpose.name }
   let(:plt_3_created)         { plate_3.created_at.strftime('%Y-%m-%d %H:%M:%S') }
 
-  let(:headers_line)          { 'Barcode,HumanBarcode,Type,Created,Location,Service,Study,Owner' }
+  let(:headers_line)          { 'Ean13Barcode,HumanBarcode,Type,Created,Location,Service,Study,Owner' }
   let(:locn_prefix)           { 'Sanger - Ogilvie - AA209 - Freezer 1' }
```
Comment (JamesGlover) [inline spec/models/location_report_spec.rb]: Probably not Ean13 here, it doesn't mean much to the users.

**Pull request merged**